prover/crates/lib/circuit_prover_service/src/gpu_circuit_prover/gpu_circuit_prover_job_saver.rs

```diff
@@ -0,0 +1,126 @@
+use std::{sync::Arc, time::Instant};
+
+use anyhow::Context;
+use async_trait::async_trait;
+use zksync_object_store::ObjectStore;
+use zksync_prover_dal::{ConnectionPool, Prover, ProverDal};
+use zksync_prover_fri_types::FriProofWrapper;
+use zksync_prover_job_processor::JobSaver;
+use zksync_types::{protocol_version::ProtocolSemanticVersion, prover_dal::FriProverJobMetadata};
+
+use crate::{gpu_circuit_prover::GpuCircuitProverExecutor, metrics::CIRCUIT_PROVER_METRICS};
+
+/// GpuCircuitProver job saver implementation.
+/// Persists the job execution to database. In case of success, artifacts are uploaded to object store.
+#[derive(Debug)]
+pub struct GpuCircuitProverJobSaver {
+    connection_pool: ConnectionPool<Prover>,
+    object_store: Arc<dyn ObjectStore>,
+    protocol_version: ProtocolSemanticVersion,
+}
+
+impl GpuCircuitProverJobSaver {
+    pub fn new(
+        connection_pool: ConnectionPool<Prover>,
+        object_store: Arc<dyn ObjectStore>,
+        protocol_version: ProtocolSemanticVersion,
+    ) -> Self {
+        Self {
+            connection_pool,
+            object_store,
+            protocol_version,
+        }
+    }
+}
+
+#[async_trait]
+impl JobSaver for GpuCircuitProverJobSaver {
+    type ExecutorType = GpuCircuitProverExecutor;
+
+    #[tracing::instrument(
+        name = "gpu_circuit_prover_job_saver",
+        skip_all,
+        fields(l1_batch = % data.1.block_number)
+    )]
+    async fn save_job_result(
+        &self,
+        data: (anyhow::Result<FriProofWrapper>, FriProverJobMetadata),
+    ) -> anyhow::Result<()> {
+        let start_time = Instant::now();
+        let (result, metadata) = data;
+        tracing::info!(
+            "Started saving gpu circuit prover job {}, on batch {}, for circuit {}, at round {}",
+            metadata.id,
+            metadata.block_number,
+            metadata.circuit_id,
+            metadata.aggregation_round
+        );
+
+        match result {
+            Ok(proof_wrapper) => {
+                let mut connection = self
+                    .connection_pool
+                    .connection()
+                    .await
+                    .context("failed to get db connection")?;
```

Comment:
it's probably very minor but is it on purpose that we establish connection before saving to the object store? It may take non-trivial amount of time, so shall we maybe only establish it when we actually need it